### PR TITLE
[Captcha] Failed login attempts

### PR DESCRIPTION
**Pull request by vincentsalucci**
## Type of change

- [X] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Code changes
- **two-factor.component:** Extended `CaptchaProtectedComponent` and applied the setup, response handling, and passing of the `captchaResponse` into the the `logInTwoFactor` call
- **(abstraction/service)auth.service:** Updated `logInTwoFactor` method to take a passed in `captchaResponse`. Updated the state clearing to allow for captcha responses triggered from failed `2fa` attempts.
- **logInStrategy.ts**: Base class - extended the `logInTwoFactor` method to allow a passed in `captchaResponse`. Created a `bypassToken` variable that is set while processing the response of a `IdentityTwoFactorResponse` and is persisted through the `login -> 2fa` flow. The variable is contained within the strategy state and is cleared once a successful login occurs. This fixes the bug introduced from the `AuthService` refactor.
- **passwordLoginStrategy.ts**: Override the `logInTwoFactor` method in order to update the request token based on the presence of either 1) a passed in `captchaResponse` or 2) a persisted `bypassToken` that was received in a previous step.
- **cli/login.command.ts**: Added the functionality to handle a captcha response during the `2fa` flow. Restructured/organized all captcha handling code into one method that returns a `tuple` of data (in order to keep the exact same flow as is currently implemented). 

## Testing requirements
- Continual regression of all things related to the `AuthService` refactor (login, 2fa, sso login, api key login, captcha)
- Specific test case: Login -> Captcha Required -> Successful Captcha -> Complete login
- Specific test case: Login -> Captcha Required -> Successful Captcha -> 2fa (incorrect/correct 2fa tokens) -> Complete login
- Specific test case: Login -> 2fa required -> Captcha Required -> Successful Captcha -> Complete login

## Dependency
[Server - Captcha failed auth attempts ceiling](https://github.com/bitwarden/server/pull/1870)

## Before you submit

- [X] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [X] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

**Review by eliykat — CHANGES_REQUESTED** [review]
Nice work! You've understood what I was going for with the authService refactor, and good pick up on not saving the `captchaToken` in the logInStrategy.

Comments below, mostly caused by this pesky `login.command.ts`...

File: node/src/cli/commands/login.command.ts [hunk 1]
```diff
@@ -435,6 +444,48 @@ export class LoginCommand {
...
+
+    try {
+      const captchaClientSecret = await this.apiClientSecret(true);
+      if (Utils.isNullOrWhitespace(captchaClientSecret)) {
+        return [null, null, badCaptcha];
+      }
+
+      let authResultResponse: AuthResult = null;
+      if (credentials != null) {
+        credentials.captchaToken = captchaClientSecret;
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply node/src/cli/commands/login.command.ts:465]: First, I definitely agree with what you're saying here! My only defense is that I **know** a future refactor of this entire command will eventually occur. My thought was to group as much as possible together in order to simplify the re-arch when it happens. I got rid of the tuple thanks to your suggestion and I feel like this is a clean solution for its purpose described above.

File: node/src/cli/commands/login.command.ts [hunk 2]
```diff
@@ -435,6 +444,48 @@ export class LoginCommand {
     }
   }
 
+  private async handleCaptchaRequired(
+    twoFactorRequest: TokenRequestTwoFactor,
+    credentials: PasswordLogInCredentials = null
+  ): Promise<[string, AuthResult, Response]> {
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply node/src/cli/commands/login.command.ts:450]: Due to another suggestion - I've simplified this a multiple return type!

File: node/src/cli/commands/login.command.ts [hunk 3]
```diff
@@ -177,39 +182,23 @@ export class LoginCommand {
...
-            "Please use your API key to validate this request and ensure BW_CLIENTSECRET is correct, if set.\n" +
-            "(https://bitwarden.com/help/cli-auth-challenges)"
-        );
-
-        try {
-          const captchaClientSecret = await this.apiClientSecret(true);
-          if (Utils.isNullOrWhitespace(captchaClientSecret)) {
-            return badCaptcha;
-          }
+        const twoFactorRequest: TokenRequestTwoFactor = {
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply node/src/cli/commands/login.command.ts:185]: Great catch - updated.

File: common/src/services/auth.service.ts [hunk 4]
```diff
@@ -115,16 +116,19 @@ export class AuthService implements AuthServiceAbstraction {
...
       throw new Error(this.i18nService.t("sessionTimeout"));
     }
 
     try {
-      const result = await this.logInStrategy.logInTwoFactor(twoFactor);
+      const result = await this.logInStrategy.logInTwoFactor(twoFactor, captchaResponse);
 
       // Only clear state if 2FA token has been accepted, otherwise we need to be able to try again
-      if (!result.requiresTwoFactor) {
+      if (!result.requiresTwoFactor && Utils.isNullOrWhitespace(result.captchaSiteKey)) {
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply common/src/services/auth.service.ts:131]: Oof - good catch. Tunnel vision on this one. Updated.

File: common/src/misc/logInStrategies/logIn.strategy.ts [hunk 5]
```diff
@@ -152,6 +156,7 @@ export abstract class LogInStrategy {
     const result = new AuthResult();
     result.twoFactorProviders = response.twoFactorProviders2;
     this.twoFactorService.setProviders(response);
+    this.bypassToken = response.captchaToken != null ? response.captchaToken : null;
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply common/src/misc/logInStrategies/logIn.strategy.ts:159]: Good point - the null coalesce operator you suggested should work fine. Updated.

File: node/src/cli/commands/login.command.ts [hunk 6]
```diff
@@ -276,11 +265,31 @@ export class LoginCommand {
...
         }
 
-        response = await this.authService.logInTwoFactor({
+        response = await this.authService.logInTwoFactor(
+          {
+            provider: selectedProvider.type,
+            token: twoFactorToken,
+            remember: false,
+          },
+          captchaClientSecret
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply node/src/cli/commands/login.command.ts:274]: Wow! Yesssss - this will help with your other comments on the `handleCaptchaRequired` method as well. Updated.

File: common/src/misc/logInStrategies/passwordLogin.strategy.ts [hunk 7]
```diff
@@ -59,6 +61,14 @@ export class PasswordLogInStrategy extends LogInStrategy {
     await this.cryptoService.setKeyHash(this.localHashedPassword);
   }
 
+  async logInTwoFactor(
+    twoFactor: TokenRequestTwoFactor,
+    captchaResponse: string
+  ): Promise<AuthResult> {
+    this.tokenRequest.captchaResponse = captchaResponse ?? this.bypassToken;
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply common/src/misc/logInStrategies/passwordLogin.strategy.ts:68]: I **believe** your understanding is correct. And while that is an extreme edge case - I guess it **is** possible. Might need to check with other parties and see if we should guard against that flow (could be a fast-follower)

File: common/src/misc/logInStrategies/logIn.strategy.ts [hunk 8]
```diff
@@ -44,7 +45,10 @@ export abstract class LogInStrategy {
     credentials: ApiLogInCredentials | PasswordLogInCredentials | SsoLogInCredentials
   ): Promise<AuthResult>;
 
-  async logInTwoFactor(twoFactor: TokenRequestTwoFactor): Promise<AuthResult> {
+  async logInTwoFactor(
+    twoFactor: TokenRequestTwoFactor,
+    captchaResponse: string
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply common/src/misc/logInStrategies/logIn.strategy.ts:50]: Great suggestions - I kept the parity by making the base class default to `null` for that value. This keeps everything upstream the same while also handling your concern of not explicitly handling a non-used value. 👍

File: node/src/cli/commands/login.command.ts [hunk 9]
```diff
@@ -69,6 +70,10 @@ export class LoginCommand {
     let clientId: string = null;
     let clientSecret: string = null;
 
+    let captchaClientSecret: string = null;
+
+    let selectedProvider: any = null;
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply node/src/cli/commands/login.command.ts:73]: Hmm - I'm a bit nervous to adjust that since its a bit outside the scope of my changes. Would feel more comfortable tackling that outside of this PR.

File: common/src/misc/logInStrategies/logIn.strategy.ts [hunk 10]
```diff
@@ -47,7 +47,7 @@ export abstract class LogInStrategy {
 
   async logInTwoFactor(
     twoFactor: TokenRequestTwoFactor,
-    captchaResponse: string
+    captchaResposne: string = null
```
Comment (eliykat) [inline common/src/misc/logInStrategies/logIn.strategy.ts:50]: Typo

File: node/src/cli/commands/login.command.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply node/src/cli/commands/login.command.ts:73]: I think it's a pretty safe change, but we can leave it for the refactor.

File: common/src/misc/logInStrategies/logIn.strategy.ts (hunk 10 above)
Comment (vincentsalucci) [reply common/src/misc/logInStrategies/logIn.strategy.ts:50]: 🤦

File: common/src/services/auth.service.ts [hunk 11]
```diff
@@ -16,6 +16,7 @@ import { KdfType } from "../enums/kdfType";
 import { ApiLogInStrategy } from "../misc/logInStrategies/apiLogin.strategy";
 import { PasswordLogInStrategy } from "../misc/logInStrategies/passwordLogin.strategy";
 import { SsoLogInStrategy } from "../misc/logInStrategies/ssoLogin.strategy";
+import { Utils } from "../misc/utils";
```
Comment (eliykat) [inline common/src/services/auth.service.ts:19]: Unused import

File: node/src/cli/commands/login.command.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply node/src/cli/commands/login.command.ts:456]: I agree simplifying the return type has made this much better. Nice work 👍 it can otherwise stay as is for now

File: common/src/services/auth.service.ts [hunk 12]
```diff
@@ -115,16 +116,19 @@ export class AuthService implements AuthServiceAbstraction {
     return result;
   }
 
-  async logInTwoFactor(twoFactor: TokenRequestTwoFactor): Promise<AuthResult> {
+  async logInTwoFactor(
+    twoFactor: TokenRequestTwoFactor,
+    captchaResponse: string
```
Comment (eliykat) [inline common/src/services/auth.service.ts:121]: I _think_ this can be removed now? Per other feedback, the captchaResponse is persisted in the logInStrategy, so we don't need to return it from `handleCaptchaRequired`. But that also means that we don't need to pass it to `authService.logInTwoFactor` or up the chain to the strategy. So I think you can revert this change, plus the changes in the logInStrategy base + sub class.

**Review by eliykat — CHANGES_REQUESTED** [review]
Sorry for the multiple reviews, Github is not keeping up with my zooming around between tabs + windows + commits. Just getting this feedback in so I don't lose it.

File: common/src/misc/logInStrategies/logIn.strategy.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply common/src/misc/logInStrategies/logIn.strategy.ts]: My concern was more that other strategy subclasses (i.e. api or SSO) appear to take a `captchaResponse` argument in this method, and the expectation would therefore be that they handle captcha as well. Instead it's basically an implicit noop, and it's mixing subclass requirements into the base class.

That said, I think this can be removed altogether, see [my other comment](https://github.com/bitwarden/jslib/pull/698#discussion_r818193106). So it may not matter.

File: common/src/services/auth.service.ts (hunk 12 above)
Comment (vincentsalucci) [reply common/src/services/auth.service.ts:121]: In the `cli` flow that is correct. However, because captcha can be triggered **on** the `two-factor` component on the other clients now, we must pass in the `captchaResponse` (the response retrieved locally from the captcha validation) which is used on the server to approve the captcha required and continue to the 2fa validation and allow login to complete.

File: node/src/cli/commands/login.command.ts [hunk 13]
```diff
@@ -69,6 +70,10 @@ export class LoginCommand {
     let clientId: string = null;
     let clientSecret: string = null;
 
+    let captchaClientSecret: string = null;
```
[the remark this answers is not in the document]
Comment (eliykat) [reply node/src/cli/commands/login.command.ts]: Good point, I missed that. And this is now a `const` in the method that requires it, beautiful.

File: common/src/services/auth.service.ts (hunk 12 above)
Comment (eliykat) [reply common/src/services/auth.service.ts:120]: Ah that's a good point

File: common/src/misc/logInStrategies/logIn.strategy.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply common/src/misc/logInStrategies/logIn.strategy.ts]: Let's tidy this up as a fast follow once we know how we're going to handle SSO + Bitwarden 2FA.

File: common/src/misc/logInStrategies/passwordLogin.strategy.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply common/src/misc/logInStrategies/passwordLogin.strategy.ts]: Agreed we'll handle it as a fast follow, please follow up at your end

**Review by eliykat — APPROVED** [review]
Great work on this!
